### PR TITLE
[Merged by Bors] - feat: lemmas about nilpotency and polynomials

### PR DESCRIPTION
**Pull request by ocfnash**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/GroupWithZero/Power.lean [hunk 1]
```diff
@@ -15,6 +15,25 @@ In this file we define integer power functions for groups with an adjoined zero
...
+section MonoidWithZero
+
+variable {M₀ : Type _} [MonoidWithZero M₀] {x y : M₀} {k : ℕ}
+
+/-- If `x` is not a (left) zero divisor then neither is `x^k` for any `k`. -/
+lemma eq_zero_of_forall_mul_eq_zero_of_mul_pow_eq_zero_left
+    (h : ∀ y, y * x = 0 → y = 0) (h' : y * (x^k) = 0) : y = 0 := by
+  induction' k with l ih; simpa using h'
+  rw [pow_succ', ← mul_assoc] at h'
+  exact ih (h _ h')
```
Comment (eric-wieser) [inline Mathlib/Algebra/GroupWithZero/Power.lean:27]: Is this meaningfully different from `LeftRegular.pow` (if that exists)? Should we have an `IsRightZeroDivisor` predicate to make this easier to parse?

File: Mathlib/Data/Polynomial/Coeff.lean [hunk 2]
```diff
@@ -300,10 +300,26 @@ theorem mul_X_pow_injective (n : ℕ) : Function.Injective fun P : R[X] => X ^ n
   rw [← coeff_X_pow_mul P n i, hPQ, coeff_X_pow_mul Q n i]
 #align polynomial.mul_X_pow_injective Polynomial.mul_X_pow_injective
 
+theorem pow_X_mul_injective (n : ℕ) : Function.Injective fun P : R[X] => P * X ^ n := by
```
Comment (eric-wieser) [inline Mathlib/Data/Polynomial/Coeff.lean:303]: This would be nice to state as `RightRegular (X^n)` (which is defeq) or similar

File: Mathlib/RingTheory/Polynomial/Nilpotent.lean [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+Copyright (c) 2023 Oliver Nash. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Emilie Uthaiwat, Oliver Nash
+-/
+import Mathlib.RingTheory.Polynomial.Basic
+
+/-!
+# Nilpotency in polynomial rings.
+
+This file is a place for results related nilpotency in (single-variable) polynomial rings.
```
Comment (adomani) [inline Mathlib/RingTheory/Polynomial/Nilpotent.lean:11]:
```suggestion
This file is a place for results related to nilpotency in (single-variable) polynomial rings.
```
Just a typo.

File: Mathlib/Algebra/GroupWithZero/Power.lean [hunk 4]
```diff
@@ -15,6 +15,25 @@ In this file we define integer power functions for groups with an adjoined zero
 This generalises the integer power function on a division ring.
 -/
 
+section MonoidWithZero
+
+variable {M₀ : Type _} [MonoidWithZero M₀] {x y : M₀} {k : ℕ}
+
+/-- If `x` is not a (left) zero divisor then neither is `x^k` for any `k`. -/
```
Comment (adomani) [inline Mathlib/Algebra/GroupWithZero/Power.lean:22]: Isn't this true for multiplication on the left by *any* element, not just powers of `x`?

File: Mathlib/Algebra/GroupWithZero/Power.lean [hunk 5]
```diff
@@ -15,6 +15,25 @@ In this file we define integer power functions for groups with an adjoined zero
...
+lemma eq_zero_of_forall_mul_eq_zero_of_mul_pow_eq_zero_left
+    (h : ∀ y, y * x = 0 → y = 0) (h' : y * (x^k) = 0) : y = 0 := by
+  induction' k with l ih; simpa using h'
+  rw [pow_succ', ← mul_assoc] at h'
+  exact ih (h _ h')
+
+/-- If `x` is not a (right) zero divisor then neither is `x^k` for any `k`. -/
```
Comment (adomani) [inline Mathlib/Algebra/GroupWithZero/Power.lean:29]: Let me try again: I think that more generally, the product of any two non right zero divisors is a non right zero divisor.  That is, the set of non right zero divisors is a semigroup.

Unless I am still confused, the proof is just associativity and application twice of the non-zero divisor hypothesis.

File: Mathlib/Data/Polynomial/Basic.lean [hunk 6]
```diff
@@ -889,6 +891,9 @@ theorem X_pow_eq_monomial (n) : X ^ n = monomial n (1 : R) := by
   · rw [pow_succ', hn, X, monomial_mul_monomial, one_mul]
 #align polynomial.X_pow_eq_monomial Polynomial.X_pow_eq_monomial
 
+@[simp] lemma coeff_pow_X {m n : ℕ} : coeff (X^n) m = if n = m then 1 else 0 := by
```
Comment (adomani) [inline Mathlib/Data/Polynomial/Basic.lean:894]: I think that `pow` and `X` should be swapped and also that `Polynomial.coeff_X_pow` exists, right?

File: Mathlib/Data/Polynomial/Coeff.lean (hunk 2 above)
Comment (adomani) [reply Mathlib/Data/Polynomial/Coeff.lean:303]: I think that `Polynomial.Monic.isRegular` exists as well.

File: Mathlib/RingTheory/Nilpotent.lean [hunk 7]
```diff
@@ -202,7 +228,20 @@ end Commute
...
-variable [CommSemiring R]
+variable [CommSemiring R] {x y : R}
+
+lemma isNilpotent_sum {ι : Type _} {s : Finset ι} {f : ι → R}
+    (hnp : ∀ i ∈ s, IsNilpotent (f i)) :
+    IsNilpotent (s.sum f) :=
+  Commute.isNilpotent_sum hnp fun _ _ _ _ ↦ Commute.all _ _
+
+@[simp] lemma isNilpotent_mul_left_iff [NoZeroDivisors R] (hy : y ≠ 0) :
+    IsNilpotent (x * y) ↔ IsNilpotent x :=
```
Comment (adomani) [inline Mathlib/RingTheory/Nilpotent.lean:239]: I'm reviewing on mobile, so I may be misunderstanding.  NoZeroDivisors implies in particular that only `0` is nilpotent, right?

File: Mathlib/Data/Polynomial/Coeff.lean (hunk 2 above)
Comment (adomani) [reply Mathlib/Data/Polynomial/Coeff.lean:303]: Looking at `Polynomial.Monic.isRegular` I think that the proof uses subtractions, whereas the result should be true for semirings.  In fact, assuming that the leading coefficient is regular should be enough.  The proof should go via Polynomial.coeff_mul_degree_add_degree.

**Comment by adomani** [comment]
This all looks good to me!  I left a few comments.

Also, thank you for dealing with the sign change on the exponents: that wrinkle was not something I was looking forward to iron out, but you found a very sneaky way to do it!

File: Mathlib/RingTheory/Polynomial/Nilpotent.lean (hunk 3 above)
Comment (ocfnash) [reply Mathlib/RingTheory/Polynomial/Nilpotent.lean:11]: Thanks!

File: Mathlib/Algebra/GroupWithZero/Power.lean (hunk 5 above)
Comment (ocfnash) [reply Mathlib/Algebra/GroupWithZero/Power.lean:29]: This is a great suggestion. I have dropped these lemmas an instead proposed introducing `nonZeroDivisorsLeft` and `nonZeroDivisorsRight` so that I can get this result via `Submonoid.pow_mem` which is essentially what you're suggesting. Thanks!

File: Mathlib/RingTheory/Nilpotent.lean (hunk 7 above)
Comment (ocfnash) [reply Mathlib/RingTheory/Nilpotent.lean:239]: Absolutely, this was a thinko on my part, caused by mechanical generalisation.

File: Mathlib/Data/Polynomial/Coeff.lean (hunk 2 above)
Comment (eric-wieser) [reply Mathlib/Data/Polynomial/Coeff.lean:303]: Just to make my suggestion be concrete
```suggestion
theorem X_pow_isRegular (n : ℕ) : IsRegular (X ^ n) := by
```
from which this result and the symmetric one is a single projection away (`.left` and ` .right`)

Comment (adomani) [reply Mathlib/Data/Polynomial/Coeff.lean]: Btw, my comment about was wrong: I confused `IsRegular` with not being a zero-divisor.  To atone, I PRed a counterexample: #6476

File: Mathlib/Algebra/Regular/Basic.lean [hunk 8]
```diff
@@ -266,6 +268,14 @@ theorem not_isRightRegular_zero [nR : Nontrivial R] : ¬IsRightRegular (0 : R) :
 theorem not_isRegular_zero [Nontrivial R] : ¬IsRegular (0 : R) := fun h => IsRegular.ne_zero h rfl
 #align not_is_regular_zero not_isRegular_zero
 
+@[simp] lemma IsLeftRegular.mul_left_eq_zero_iff (hb : IsLeftRegular b) : b * a = 0 ↔ a = 0 := by
+  nth_rw 1 [← mul_zero b]
+  exact ⟨fun h ↦ hb h, fun ha ↦ by rw [ha, mul_zero]⟩
```
Comment (adomani) [inline Mathlib/Algebra/Regular/Basic.lean:273]:
```suggestion
  exact ⟨hb, fun ha ↦ by rw [ha, mul_zero]⟩
```
Does this work?

File: Mathlib/GroupTheory/Submonoid/ZeroDivisors.lean [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+## Main definitions
+
+ * `nonZeroDivisorsLeft`: the elements of a `MonoidWithZero` that are not left zero divisors.
+ * `nonZeroDivisorsRight`: the elements of a `MonoidWithZero` that are not right zero divisors.
+
+-/
+
+section MonoidWithZero
+
+variable (M₀ : Type _) [MonoidWithZero M₀]
```
Comment (adomani) [inline Mathlib/GroupTheory/Submonoid/ZeroDivisors.lean:20]: Technically, this probably makes sense for `SemigroupWithZero` and `Subsemigroup` as well...

**Review by adomani — COMMENTED** [review]
A couple of extra really minor comments!

File: Mathlib/Algebra/Regular/Basic.lean (hunk 8 above)
Comment (ocfnash) [reply Mathlib/Algebra/Regular/Basic.lean:273]: I like the idea but in this case I think the suggestion makes this slightly harder to read.

File: Mathlib/GroupTheory/Submonoid/ZeroDivisors.lean (hunk 9 above)
Comment (ocfnash) [reply Mathlib/GroupTheory/Submonoid/ZeroDivisors.lean:20]: It does but I need the `Submonoid` type since I want to use `Submonoid.pow_mem` (which cannot exist for `Subsemigroup` because the power might be zero).

We could of course make both definitions but I propose waiting since we do not have an application in mind.

Comment (adomani) [reply Mathlib/GroupTheory/Submonoid/ZeroDivisors.lean:20]: Sure, it just got me thinking about whether I had ever encountered a ring without unit and with nilpotent elements, but I could not come up with one.  Likely, because I do not often see rings without units!  :smile:

**Comment by adomani** [comment]
This looks good to me!  Thanks!

File: Mathlib/Data/Polynomial/Coeff.lean [hunk 10]
```diff
@@ -293,16 +293,16 @@ theorem mul_X_pow_eq_zero {p : R[X]} {n : ℕ} (H : p * X ^ n = 0) : p = 0 :=
   ext fun k => (coeff_mul_X_pow p n k).symm.trans <| ext_iff.1 H (k + n)
 #align polynomial.mul_X_pow_eq_zero Polynomial.mul_X_pow_eq_zero
 
-theorem mul_X_pow_injective (n : ℕ) : Function.Injective fun P : R[X] => X ^ n * P := by
+-- TODO Unify this with `Polynomial.Monic.isRegular`
+theorem isRegular_X_pow (n : ℕ) : IsRegular ((X : R[X]) ^ n) := by
```
Comment (eric-wieser) [inline Mathlib/Data/Polynomial/Coeff.lean:297]: nit (untested):
```suggestion
theorem isRegular_X_pow (n : ℕ) : IsRegular (X ^ n : R[X]) := by
```

File: Mathlib/Data/Polynomial/Coeff.lean [hunk 11]
```diff
@@ -293,16 +293,16 @@ theorem mul_X_pow_eq_zero {p : R[X]} {n : ℕ} (H : p * X ^ n = 0) : p = 0 :=
...
   intro P Q hPQ
   simp only at hPQ
   ext i
   rw [← coeff_X_pow_mul P n i, hPQ, coeff_X_pow_mul Q n i]
-#align polynomial.mul_X_pow_injective Polynomial.mul_X_pow_injective
 
-theorem mul_X_injective : Function.Injective fun P : R[X] => X * P :=
-  pow_one (X : R[X]) ▸ mul_X_pow_injective 1
-#align polynomial.mul_X_injective Polynomial.mul_X_injective
+@[simp] theorem isRegular_X : IsRegular (X : R[X]) := by simpa using isRegular_X_pow 1
```
Comment (eric-wieser) [inline Mathlib/Data/Polynomial/Coeff.lean:305]: I guess you can probably prove these in the other direction and use `isRegular_X.pow n`?

File: Mathlib/RingTheory/Nilpotent.lean [hunk 12]
```diff
@@ -202,7 +229,12 @@ end Commute
 
 section CommSemiring
 
-variable [CommSemiring R]
+variable [CommSemiring R] {x y : R}
+
+lemma isNilpotent_sum {ι : Type _} {s : Finset ι} {f : ι → R}
+    (hnp : ∀ i ∈ s, IsNilpotent (f i)) :
+    IsNilpotent (s.sum f) :=
```
Comment (eric-wieser) [inline Mathlib/RingTheory/Nilpotent.lean:236]: nit: big operator notation?

File: Mathlib/Data/Polynomial/Coeff.lean (hunk 10 above)
Comment (adomani) [reply Mathlib/Data/Polynomial/Coeff.lean:297]: While we are on the topic, @eric-wieser, what do you think of `(X (R := R) ^ n)`?

Comment (eric-wieser) [reply Mathlib/Data/Polynomial/Coeff.lean:297]: I think that's worse

File: Mathlib/RingTheory/Nilpotent.lean (hunk 12 above)
Comment (ocfnash) [reply Mathlib/RingTheory/Nilpotent.lean]: Thanks, done.

File: Mathlib/Data/Polynomial/Coeff.lean (hunk 11 above)
Comment (ocfnash) [reply Mathlib/Data/Polynomial/Coeff.lean]: OK I've made this change but it doesn't really affect anything because of the existing API for polynomials.

File: Mathlib/Data/Polynomial/Coeff.lean (hunk 10 above)
Comment (ocfnash) [reply Mathlib/Data/Polynomial/Coeff.lean]: I agree with Eric here and have made this suggestion.

File: Mathlib/Data/Polynomial/Coeff.lean (hunk 11 above)
Comment (ocfnash) [reply Mathlib/Data/Polynomial/Coeff.lean]: Update: looking at the diff now, I do prefer the new version (using `IsRegular.pow`).

File: Mathlib/Data/Polynomial/Coeff.lean [hunk 13]
```diff
@@ -293,16 +294,15 @@ theorem mul_X_pow_eq_zero {p : R[X]} {n : ℕ} (H : p * X ^ n = 0) : p = 0 :=
...
   ext i
-  rw [← coeff_X_pow_mul P n i, hPQ, coeff_X_pow_mul Q n i]
-#align polynomial.mul_X_pow_injective Polynomial.mul_X_pow_injective
+  rw [← coeff_X_mul P i, hPQ, coeff_X_mul Q i]
 
-theorem mul_X_injective : Function.Injective fun P : R[X] => X * P :=
-  pow_one (X : R[X]) ▸ mul_X_pow_injective 1
-#align polynomial.mul_X_injective Polynomial.mul_X_injective
+-- TODO Unify this with `Polynomial.Monic.isRegular`
+theorem isRegular_X_pow (n : ℕ) : IsRegular ((X ^ n : R[X])) := isRegular_X.pow n
```
Comment (eric-wieser) [inline Mathlib/Data/Polynomial/Coeff.lean:305]:
```suggestion
theorem isRegular_X_pow (n : ℕ) : IsRegular (X ^ n : R[X]) := isRegular_X.pow n
```

File: Mathlib/GroupTheory/Submonoid/ZeroDivisors.lean [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+
+@[simp] lemma coe_nonZeroDivisorsRight_eq [NoZeroDivisors M₀] [Nontrivial M₀] :
+    nonZeroDivisorsRight M₀ = {x : M₀ | x ≠ 0} := by
+  ext x
+  simp only [SetLike.mem_coe, mem_nonZeroDivisorsRight_iff, mul_eq_zero, Set.mem_setOf_eq]
+  refine' ⟨fun h ↦ _, fun hx y hx' ↦ by aesop⟩
+  contrapose! h
+  exact ⟨1, Or.inl h, one_ne_zero⟩
+
+end MonoidWithZero
```
Comment (eric-wieser) [inline Mathlib/GroupTheory/Submonoid/ZeroDivisors.lean:61]: Could you add the lemma that the two submonoids are equal in commutative rings?

File: Mathlib/Order/SuccPred/Basic.lean [hunk 15]
```diff
@@ -1511,3 +1511,14 @@ theorem Pred.rec_top (p : α → Prop) (htop : p ⊤) (hpred : ∀ a, p a → p
 #align pred.rec_top Pred.rec_top
 
 end OrderTop
+
+lemma SuccOrder.forall_ne_bot_iff
+    {α : Type _} [Nontrivial α] [PartialOrder α] [OrderBot α] [SuccOrder α] [IsSuccArchimedean α]
```
Comment (eric-wieser) [inline Mathlib/Order/SuccPred/Basic.lean:1516]: Isn't there already an `α` around?

**Review by eric-wieser — APPROVED** [review]
bors d+

**Comment by ocfnash** [comment]
bors merge

File: Mathlib/Order/SuccPred/Basic.lean [hunk 16]
```diff
@@ -1511,3 +1511,14 @@ theorem Pred.rec_top (p : α → Prop) (htop : p ⊤) (hpred : ∀ a, p a → p
 #align pred.rec_top Pred.rec_top
 
 end OrderTop
+
+lemma SuccOrder.forall_ne_bot_iff
```
Comment (eric-wieser) [inline Mathlib/Order/SuccPred/Basic.lean:1515]: Can you add the `PredOrder` version too, or is it not trivially true via the OrderDual?
